tests/static-analysis/associate.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/**
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+include __DIR__ . '/../../vendor/autoload.php';
+
+use loophp\collection\Collection;
+
+/**
+ * @param Collection<int, int> $collection
+ */
+function associate_checkNumeric(Collection $collection): void
+{
+}
+/**
+ * @param Collection<string, int> $collection
+ */
+function associate_checkMap(Collection $collection): void
+{
+}
+/**
+ * @param Collection<int, int|string> $collection
+ */
+function associate_checkMixed(Collection $collection): void
+{
+}
+
+associate_checkNumeric(Collection::empty()->associate());
+associate_checkMap(Collection::empty()->associate());
+associate_checkMixed(Collection::empty()->associate());
+
+associate_checkNumeric(Collection::fromIterable([1, 2, 3])->associate());
+associate_checkMap(Collection::fromIterable(['foo' => 1, 'bar' => 2])->associate());
+associate_checkMixed(Collection::fromIterable([1, 2, 'b', '5', 4])->associate());
+
+associate_checkNumeric(Collection::fromIterable([4, 5, 6])->associate(static fn (int $carry, int $key): int => $key * 2));
+associate_checkNumeric(Collection::fromIterable([4, 5, 6])->associate(static fn (int $carry, int $key, int $value) => $key * 2));
+associate_checkNumeric(
+    Collection::fromIterable([4, 5, 6])->associate(
+    null,
+    static fn (int $carry, int $key, int $value) => $value
+)
+);
+associate_checkNumeric(Collection::fromIterable([4, 5, 6])->associate(
+    static fn (int $carry, int $key): int => $key * 2,
+    static fn (int $carry, int $key, int $value): int => $value * -1
+));
+
+associate_checkMap(Collection::fromIterable(['foo' => 1, 'bar' => 2])->associate(static fn (string $carry, string $key): string => "{$key}z"));
```

Comment:
some of the errors here are ridiculous 😄 

`ERROR: InvalidArgument - tests/static-analysis/associate.php:54:20 - Argument 1 of associate_checkMap expects loophp\collection\Collection<string, int>, loophp\collection\Collection<non-empty-string, 1|2> provided (see https://psalm.dev/004)`

Can't tell if it's an issue with the type annotations or just pushing the limits of what Psalm is capable of right now

Reply:
It goes far indeed, but I don't know what's the best option here.